### PR TITLE
Batch writes originating from logins/logouts for persistent sessions

### LINKED ISSUE #28862: Improve persistent sessions DB throughput for logins/logouts by batching
### Description

As a follow-up to #28501, use batching for also logins/logouts. As an enhancement to the previous ticket, instead of firing off the tasks and don't wait for their completion, for logins/logouts we wait for their completion to ensure their data reaches both the database, and all caches before returning to the user. 

### Discussion

_No response_

### Motivation

Currently we have one DB transaction per login and logout, leading to a lot of IOPS on the database. When the load increases, the IOPS increase linearly.

With the batching, the number of IOPS will after an initial increase stay constant as the number of commits reaches its maximum and the size of the batches increases to the current maximum of 100. 

While this is the way it works for the database, it will not work as ideally for writing to Infinispan (especially the other site's infinispan), as those writes do not benefit from the bundling.

Example (ideal world): With the time to complete a batch in 100 ms, and 200 logins per second, there will be up to 10 batches processed per second, with 20 items in each batch. 

While this optimizes the throughput especially in high load environments, we will see the average and tail latencies increase in the order of the time it needs to write a batch. With small systems with a small number of logins/logouts/refreshes per second, the added latency and tail latency will be minimal, and it will increase with the load. 

### Details

I'll prepare a PR.

### PR DESCRIPTION
**Pull request by ahus1**
Closes #28862

### CODE REVIEW

**Comment by ahus1** [comment]
I saw some deadlocks when load-testing this. Will provide an update, soon.

**Comment by ahus1** [comment]
@mhajas - I updated this PR with some changes. Now it runs nicely on a single-site Keycloak with a multi-AZ Aurora. 

It will batch the updates to the the database in the background thread, but will then use the regular task to update the Infinispan (internal and external). 

This shows the stable activities running that setup with 300 logins per second: 

image

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/JpaChangesPerformer.java
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/JpaChangesPerformer.java:73]: This method can be removed. I can do it in the follow-up PR if we don't want to wait for CI again

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java:113]: 0 means we add no backoff, is it correct? If there are more nodes in two DCs fighting to update the same entity 0 can make it longer. Maybe we can keep this 0 by default but make it configurable so we can test it within A/A perf tests?

**Review by mhajas — APPROVED** [review]
Nice work @ahus1! I like the change. I added two minor comments but I am ok to merge as is if you don't want to include any change in this PR, so approving.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/JpaChangesPerformer.java
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/JpaChangesPerformer.java:73]: Thank you for spotting this. Yes, please do it in the follow-up PR.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java:113]: In the recent tests when every update was going through the background thread, I saw no deadlocks any more. I would assume that if two nodes fight, one of them would win and it would be ok to retry immediately, so I chose no back-off period. Maybe we would just keep an eye on it and make it configurable when we see the need in one of our performance tests?

**Comment by ahus1** [comment]
Statistics for the matter of completeness: It ran nicely with 250 logins & logouts per second on the 2-AZ Aurora PostgreSQL db.t4g.large servers, and 300 seems to be the limit. 

image

**Pull request merged**
